File: 프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.py
from collections import Counter
def solution(weights):
    result = 0
    ratios = [1/2, 2/3, 3/4]
    count_dict = Counter(weights)
    weights = list(set(weights))
    
    for k, v in count_dict.items():
        if v > 1:
            result += v * (v-1) / 2
    
    for w in weights:
        for r in ratios:
            combi_num = count_dict[w * r]
            result += combi_num * count_dict[w]
    return result

File: 프로그래머스/2/152996. 시소 짝꿍/시소 짝꿍.js
function solution(weights) {
  let result = 0;
  const ratios = [1 / 2, 2 / 3, 3 / 4];

  const countMap = new Map();
  for (const w of weights) {
    countMap.set(w, (countMap.get(w) || 0) + 1);
  }

  const uniqueWeights = [...new Set(weights)];

  for (const [w, count] of countMap.entries()) {
    if (count > 1) {
      result += (count * (count - 1)) / 2;
    }
  }

  for (const w of uniqueWeights) {
    for (const r of ratios) {
      const target = w * r;
      if (countMap.has(target)) {
        result += countMap.get(target) * countMap.get(w);
      }
    }
  }

  return result;
}